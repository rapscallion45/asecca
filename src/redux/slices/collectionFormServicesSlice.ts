import {
  ActionReducerMapBuilder,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';
import {
  ICollectionFormServicesContactsDataPayload,
  ICollectionFormServicesData,
  ICollectionFormServicesDataPayload,
} from '@/lib/api/api-types';
import collectionFormService from '../../services/forms/collectionFormService';
import {
  ICollectionFormServicesState,
  IFetchCollectionFormServicesByCollectionIdArgs,
  ISaveCollectionFormServicesByCollectionIdArgs,
  ICollectionFormServicesEditSchedulePayload,
} from '../types';
import { addNotification } from './notificationsSlice';
import collectionFormServicesDataMock from '../../../__mocks__/collectionFormServicesDataMock';
import collectionFormServicesContactsDataMock from '../../../__mocks__/collectionFormServicesContactsDataMock';

/**
 * State slice definition for Collection Form Services
 *
 * All state updates to the Collection Form Services are managed here.
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.14
 */

/**
 * Async thunk for GET /api/collection/services/api/get API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.14
 * @memberof collectionFormServicesReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchByCollectionId = createAsyncThunk(
  'collectionFormServices/fetchByCollectionId',
  async (args: IFetchCollectionFormServicesByCollectionIdArgs, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getServices(args.collectionId);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Collection Form Services from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for POST /api/collection/services/api/set API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.14
 * @memberof collectionFormServicesReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const saveByCollectionId = createAsyncThunk(
  'collectionFormServices/saveByCollectionId',
  async (args: ISaveCollectionFormServicesByCollectionIdArgs, thunkAPI) => {
    const res = await collectionFormService.setServices(args.data);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to save Collection Form Services to server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    } else {
      thunkAPI.dispatch(
        addNotification({
          message: 'Successfully saved Collection Form Services to server',
          variant: 'success',
        })
      );
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for GET /api/collection/service/api/contacts API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormServicesReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchTypes = createAsyncThunk(
  'collectionFormServices/fetchTypes',
  async (args, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getServicesContacts();

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Services Contacts data from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Initialises Collection Form Services state to empty array
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.14
 * @memberof collectionFormServicesReduxSlice
 *
 * @constant
 * @type {IcollectionFormServicesState}
 */
const initialcollectionFormServicesState: ICollectionFormServicesState = {
  loading: false,
  data: collectionFormServicesDataMock,
  dataShadow: collectionFormServicesDataMock,
  saving: false,
  edited: false,
  loadingContacts: false,
  contacts: collectionFormServicesContactsDataMock,
};

/**
 * Create the redux slice for interacting with the Collection Form Services state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.14
 * @memberof collectionFormServicesReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {Slice<Object>}
 */
const collectionFormServicesSlice = createSlice({
  name: 'collectionFormServices',
  initialState: initialcollectionFormServicesState,
  reducers: {
    /* reducer used for user input changes to the Services data */
    editServices: (
      state,
      action: PayloadAction<ICollectionFormServicesEditSchedulePayload>
    ) => {
      /* update schedule */
      state.data = {
        ...state.data,
        [action.payload.itemKey as keyof ICollectionFormServicesData]:
          action.payload.value,
      };
      state.edited = true;
    },
    /* reducer used for when user clears edits to Services data */
    resetServices: (state) => {
      /* reset the data by simply copying the shadow to working copy */
      state.data = state.dataShadow;
      state.edited = false;
    },
  },
  extraReducers: (
    builder: ActionReducerMapBuilder<ICollectionFormServicesState>
  ) => {
    /*
     * `extraReducers` used to handle actions that were generated
     * outside of the Collection Form Services slice, such as the async thunks
     * for the API requests, defined at the top of this file.
     */
    builder
      /* Fetch Collection Form Services extra reducers */
      .addCase(
        fetchByCollectionId.pending,
        (state: ICollectionFormServicesState) => {
          state.loading = true;
          state.data = collectionFormServicesDataMock;
          state.dataShadow = collectionFormServicesDataMock;
          state.error = undefined;
          state.edited = false;
        }
      )
      .addCase(
        fetchByCollectionId.fulfilled,
        (
          state: ICollectionFormServicesState,
          action: PayloadAction<ICollectionFormServicesDataPayload>
        ) => {
          state.loading = false;
          state.data = action.payload;
          state.dataShadow = action.payload;
          state.error = undefined;
        }
      )
      .addCase(
        fetchByCollectionId.rejected,
        (state: ICollectionFormServicesState) => {
          state.loading = false;
          state.data = collectionFormServicesDataMock;
          state.dataShadow = collectionFormServicesDataMock;
          state.error =
            'Failed to load Collection Form Services data from server.';
        }
      )
      /* Save Collection Form Services extra reducers */
      .addCase(
        saveByCollectionId.pending,
        (state: ICollectionFormServicesState) => {
          state.saving = true;
        }
      )
      .addCase(
        saveByCollectionId.fulfilled,
        (state: ICollectionFormServicesState) => {
          state.saving = false;
          state.edited = false;
        }
      )
      .addCase(
        saveByCollectionId.rejected,
        (state: ICollectionFormServicesState) => {
          state.saving = false;
        }
      )
      /* Fetch Collection Form Services Types extra reducers */
      .addCase(fetchTypes.pending, (state: ICollectionFormServicesState) => {
        state.loadingContacts = true;
      })
      .addCase(
        fetchTypes.fulfilled,
        (
          state: ICollectionFormServicesState,
          action: PayloadAction<ICollectionFormServicesContactsDataPayload>
        ) => {
          state.loadingContacts = false;
          state.contacts = action.payload;
        }
      )
      .addCase(fetchTypes.rejected, (state: ICollectionFormServicesState) => {
        state.loadingContacts = false;
        state.contacts = collectionFormServicesContactsDataMock;
        state.error =
          'Failed to load Collection Form Services Contacts data from server.';
      });
  },
});

/* Collection Form Services actions for editing and resetting schedule data */
export const { editServices, resetServices } =
  collectionFormServicesSlice.actions;

export default collectionFormServicesSlice.reducer;
