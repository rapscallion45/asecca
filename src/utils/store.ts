/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
