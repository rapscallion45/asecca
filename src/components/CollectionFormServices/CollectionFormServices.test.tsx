import React from 'react';
import { Provider } from 'react-redux';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { fetchByCollectionId as fetchCollectionFormServicesByCollectionId } from '@/redux/slices/collectionFormServicesSlice';
import store from '@/redux/store';
import {
  getCollectionFormServicesDecommissionRequestValue,
  getCollectionFormServicesOwnershipRetentionValue,
  getCollectionFormServicesRedeliveryRequestValue,
  getCollectionFormServicesTypeEnum,
} from '@/utils';
import CollectionFormServices from './CollectionFormServices';
import collectionFormServicesDataMock from '../../../__mocks__/CollectionForm/collectionFormServicesDataMock';
import collectionFormServicesContactsDataMock from '../../../__mocks__/CollectionForm/collectionFormServicesContactsDataMock';

/* default test query ID */
const query: string = '66135000001760012';

/**
 * Collection Form Services Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Collection Form Services', () => {
  it('Renders correctly', async () => {
    /** Arrange */
    store.dispatch(
      fetchCollectionFormServicesByCollectionId({
        collectionId: query,
      })
    );

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <CollectionFormServices collectionId={query} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Form Fields', () => {
    it('Should render all required New Contact data fields', async () => {
      /** Arrange */
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormServices collectionId={query} />
        </Provider>
      );

      /** Assert - ensure all fields are present */
      await waitFor(() => {
        expect(screen.getByText('Processing:')).toBeInTheDocument();
      });
      expect(screen.getByText('Service Type:')).toBeInTheDocument();
      expect(screen.getByText('Decomissioning Requested:')).toBeInTheDocument();
      expect(
        screen.getByText('Customer Retains Ownership Of Devices:')
      ).toBeInTheDocument();
      expect(screen.getByText('Redelivery Requested:')).toBeInTheDocument();
      expect(screen.getByText('Site Contact:')).toBeInTheDocument();
      expect(
        screen.getByDisplayValue(
          collectionFormServicesDataMock.on_site_processing
            ? 'On-Site'
            : 'Off-Site'
        )
      ).toBeInTheDocument();
      expect(
        screen.getByDisplayValue(
          getCollectionFormServicesTypeEnum(
            collectionFormServicesDataMock
          ) as string
        )
      ).toBeInTheDocument();
      if (
        getCollectionFormServicesDecommissionRequestValue(
          collectionFormServicesDataMock
        )
      ) {
        expect(screen.getByTestId('decomissioning-checkbox')).toBeChecked();
      } else {
        expect(screen.getByTestId('decomissioning-checkbox')).not.toBeChecked();
      }
      if (
        getCollectionFormServicesOwnershipRetentionValue(
          collectionFormServicesDataMock
        )
      ) {
        expect(screen.getByTestId('redelivery-checkbox')).toBeChecked();
      } else {
        expect(screen.getByTestId('redelivery-checkbox')).not.toBeChecked();
      }
      if (
        getCollectionFormServicesRedeliveryRequestValue(
          collectionFormServicesDataMock,
          getCollectionFormServicesOwnershipRetentionValue(
            collectionFormServicesDataMock
          )
        )
      ) {
        expect(screen.getByTestId('ownership-checkbox')).toBeChecked();
      } else {
        expect(screen.getByTestId('ownership-checkbox')).not.toBeChecked();
      }
      expect(
        screen.getByDisplayValue(collectionFormServicesDataMock.site_contact)
      ).toBeInTheDocument();
    });
  });

  describe('Form Callbacks', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call API with inputted values when values updated and save button clicked', async () => {
      /** Arrange */
      const testProcessing: string = 'On-Site';
      const testServiceType: string = 'Recycling';
      const testDecomissionRequested: boolean = true;
      const testRetainsOwnership: boolean = true;
      const testRedeliveryRequested: boolean = true;
      const testContact: string =
        collectionFormServicesContactsDataMock.contacts_list[1];

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormServices collectionId={query} />
        </Provider>
      );
      await waitFor(() => {
        expect(screen.getByText('Processing:')).toBeInTheDocument();
      });
      expect(
        screen.getByDisplayValue(collectionFormServicesDataMock.site_contact)
      ).toBeInTheDocument();
      /** Act - update field values and click Save button */
      act(() => {
        fireEvent(
          screen.getByDisplayValue(
            collectionFormServicesDataMock.on_site_processing
              ? 'On-Site'
              : 'Off-Site'
          ),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormServicesDataMock.on_site_processing
              ? 'On-Site'
              : 'Off-Site'
          ),
          {
            target: { value: testProcessing },
          }
        );
      });
      act(() => {
        fireEvent(
          screen.getByDisplayValue(
            getCollectionFormServicesTypeEnum(
              collectionFormServicesDataMock
            ) as string
          ),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
        fireEvent.change(
          screen.getByDisplayValue(
            getCollectionFormServicesTypeEnum(
              collectionFormServicesDataMock
            ) as string
          ),
          {
            target: { value: testServiceType },
          }
        );
      });
      if (testDecomissionRequested)
        act(() => {
          fireEvent(
            screen.getByTestId('decomissioning-checkbox'),
            new MouseEvent('click', {
              bubbles: true,
              cancelable: true,
            })
          );
        });
      if (testRetainsOwnership)
        act(() => {
          fireEvent(
            screen.getByTestId('ownership-checkbox'),
            new MouseEvent('click', {
              bubbles: true,
              cancelable: true,
            })
          );
        });
      if (testRedeliveryRequested)
        act(() => {
          fireEvent(
            screen.getByTestId('redelivery-checkbox'),
            new MouseEvent('click', {
              bubbles: true,
              cancelable: true,
            })
          );
        });
      act(() => {
        fireEvent(
          screen.getByDisplayValue(collectionFormServicesDataMock.site_contact),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
        fireEvent.change(
          screen.getByDisplayValue(collectionFormServicesDataMock.site_contact),
          {
            target: { value: testContact },
          }
        );
      });
      /** Assert - ensure updates have been processed */
      await waitFor(() => {
        expect(screen.getByDisplayValue(testProcessing)).toBeInTheDocument();
      });
      expect(screen.getByDisplayValue(testServiceType)).toBeInTheDocument();
      expect(screen.getByDisplayValue(testContact)).toBeInTheDocument();
      /** Act - click Commit button */
      act(() => {
        fireEvent(
          screen.getByText('Commit'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - save API to have been called with data from state */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(1);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        '/api/collection/services/api/set',
        expect.objectContaining({
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            collection: query,
            on_site_processing: testProcessing === 'On-Site',
            service_type: {
              Recycling: {
                ownership_retention: {
                  RetainsOwnership: {
                    redelivery_requested: testRedeliveryRequested,
                  },
                },
                decommissioning_requested: testDecomissionRequested,
              },
            },
            site_contact: testContact,
          }),
        })
      );
    });
  });
});
