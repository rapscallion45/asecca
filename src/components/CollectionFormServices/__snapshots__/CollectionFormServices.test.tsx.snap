// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Form Services Renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
      >
        Services
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-1hizqsf-MuiCardContent-root"
  >
    <form
      autoComplete="off"
      className="MuiBox-root css-1b7bzod"
      noValidate={true}
    >
      <div
        className="MuiBox-root css-1kukc4r"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1h1t5kx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-1unik2p-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Processing:
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <div
              className="MuiFormControl-root css-4xtla9-MuiFormControl-root"
            >
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl  css-es1mxv-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                onClick={[Function]}
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-label="collection-form-services-select"
                  aria-labelledby="collection-form-services-select"
                  className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="collection-form-services-select"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  Off-Site
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  required={false}
                  tabIndex={-1}
                  value="Off-Site"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden={true}
                  className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="css-ihdtdm"
                  >
                    <span
                      className="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root css-1kukc4r"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1h1t5kx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-1unik2p-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Service Type:
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <div
              className="MuiFormControl-root css-4xtla9-MuiFormControl-root"
            >
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl  css-es1mxv-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                onClick={[Function]}
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-label="collection-form-services-select"
                  aria-labelledby="collection-form-services-select"
                  className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="collection-form-services-select"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                   Recycling
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  required={false}
                  tabIndex={-1}
                  value="Recycling"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden={true}
                  className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="css-ihdtdm"
                  >
                    <span
                      className="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root css-1kukc4r"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1h1t5kx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-6 css-1osj8n2-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Decomissioning Requested:
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <span
              className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-root MuiCheckbox-colorSecondary css-1sg8fd1-MuiButtonBase-root-MuiCheckbox-root"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate={false}
                data-testid="decomissioning-checkbox"
                disabled={false}
                onChange={[Function]}
                required={false}
                type="checkbox"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root css-1kukc4r"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1h1t5kx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Customer Retains Ownership Of Devices:
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <span
              className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-root MuiCheckbox-colorSecondary css-1sg8fd1-MuiButtonBase-root-MuiCheckbox-root"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate={false}
                data-testid="ownership-checkbox"
                disabled={false}
                onChange={[Function]}
                required={false}
                type="checkbox"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root css-1kukc4r"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1h1t5kx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-6 css-1osj8n2-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Redelivery Requested:
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <span
              className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-root MuiCheckbox-colorSecondary css-1sg8fd1-MuiButtonBase-root-MuiCheckbox-root"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate={false}
                data-testid="redelivery-checkbox"
                disabled={false}
                onChange={[Function]}
                required={false}
                type="checkbox"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root css-1kukc4r"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1h1t5kx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-1unik2p-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Site Contact:
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <div
              className="MuiFormControl-root css-4xtla9-MuiFormControl-root"
            >
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-formControl  css-es1mxv-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                onClick={[Function]}
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-label="collection-form-services-select"
                  aria-labelledby="collection-form-services-select"
                  className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="collection-form-services-select"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  Contact One
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  required={false}
                  tabIndex={-1}
                  value="Contact One"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden={true}
                  className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="css-ihdtdm"
                  >
                    <span
                      className="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
    <div
      className="MuiBox-root css-kh5ii8"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1ex1hsz-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Cancel
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fddvdp-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        disabled={true}
        id=":r0:"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Commit
      </button>
    </div>
  </div>
</div>
`;
